controllers/control/k8ssandratask_controller.go

```diff
@@ -0,0 +1,384 @@
+				dcs = append([]k8capi.CassandraDatacenterTemplate{dc}, dcs...)
+				found = true
+				break
+			}
+		}
+		if !found {
+			unknownDcNames = append(unknownDcNames, dcName)
+		}
+	}
+	if len(unknownDcNames) > 0 {
+		return nil, fmt.Errorf("unknown datacenters: %s", strings.Join(unknownDcNames, ", "))
+	}
+	return dcs, nil
+}
+
+func newCassandraTask(kTask *api.K8ssandraTask, namespace string, dcName string) *cassapi.CassandraTask {
+	cTemplate := kTask.Spec.Template
+
+	// Never set the TTL on the CassandraTask. We manage it ourselves on the K8ssandraTask, once it expires we'll
+	// cascade-delete the dependent CassandraTasks.
+	cTemplate.TTLSecondsAfterFinished = &noTTL
+
+	return &cassapi.CassandraTask{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:   namespace,
+			Name:        cassandraTaskName(kTask, dcName),
+			Annotations: map[string]string{},
+			Labels: map[string]string{
+				k8capi.NameLabel:                k8capi.NameLabelValue,
+				k8capi.PartOfLabel:              k8capi.PartOfLabelValue,
+				k8capi.ComponentLabel:           k8capi.ComponentLabelValueCassandra,
+				k8capi.CreatedByLabel:           k8capi.CreatedByLabelValueK8ssandraTaskController,
+				api.K8ssandraTaskNameLabel:      kTask.Name,
+				api.K8ssandraTaskNamespaceLabel: kTask.Namespace,
+			},
+		},
+		Spec: cassapi.CassandraTaskSpec{
+			Datacenter:            corev1.ObjectReference{Namespace: namespace, Name: dcName},
+			CassandraTaskTemplate: cTemplate,
+		},
+	}
+}
+
+func cassandraTaskName(kTask *api.K8ssandraTask, dcName string) string {
+	return kTask.Name + "-" + dcName
+}
+
+func calculateDeletionTime(kTask *api.K8ssandraTask) time.Time {
+	if kTask.Spec.Template.TTLSecondsAfterFinished != nil {
+		if *kTask.Spec.Template.TTLSecondsAfterFinished == 0 {
+			return time.Time{}
+		}
+		return kTask.Status.CompletionTime.Add(time.Duration(*kTask.Spec.Template.TTLSecondsAfterFinished) * time.Second)
+	}
+	return kTask.Status.CompletionTime.Add(defaultTTL)
+}
+
+func (r *K8ssandraTaskReconciler) recordCassandraTaskCreated(
+	kTask *api.K8ssandraTask,
+	cTask *cassapi.CassandraTask,
+	k8sContext string,
+) {
+	contextInfo := ""
+	if k8sContext != "" {
+		contextInfo = " in context " + k8sContext
+	}
+	r.Recorder.Event(kTask, "Normal", "CreateCassandraTask",
+		fmt.Sprintf("Created CassandraTask %s.%s%s", cTask.Namespace, cTask.Name, contextInfo))
+}
+
+// reportInvalidSpec is called when the user provided an invalid K8ssandraTask spec. A warning event is emitted and the
+// task's status is set to "Invalid".
+func (r *K8ssandraTaskReconciler) reportInvalidSpec(
+	ctx context.Context,
+	kTask *api.K8ssandraTask,
+	format string,
+	arguments ...interface{},
+) (ctrl.Result, error) {
+	r.Recorder.Event(kTask, "Warning", "InvalidSpec", fmt.Sprintf(format, arguments...))
+
+	patch := client.MergeFrom(kTask.DeepCopy())
+	kTask.SetCondition(api.JobInvalid, corev1.ConditionTrue)
+	err := r.Status().Patch(ctx, kTask, patch)
+	return ctrl.Result{}, err
```

Comment:
Is the requeue intentional after setting it to Invalid? Wouldn't it be enough to reprocess it when it has been changed?